builtin/logical/pki/path_tidy.go

```diff
@@ -440,6 +460,114 @@ func (b *backend) doTidyRevocationStore(ctx context.Context, req *logical.Reques
 	return nil
 }
 
+func (b *backend) doTidyExpiredIssuers(ctx context.Context, req *logical.Request, logger hclog.Logger, config *tidyConfig) error {
+	if b.System().ReplicationState().HasState(consts.ReplicationDRSecondary|consts.ReplicationPerformanceStandby) ||
+		(!b.System().LocalMount() && b.System().ReplicationState().HasState(consts.ReplicationPerformanceSecondary)) {
+		b.Logger().Debug("skipping expired issuer tidy as we're not on the primary or secondary with a local mount")
+		return nil
+	}
+
+	// Short-circuit to avoid having to deal with the legacy mounts. While we
+	// could handle this case and remove these issuers, its somewhat
+	// unexpected behavior and we'd prefer to finish the migration first.
+	if b.useLegacyBundleCaStorage() {
+		return nil
+	}
+
+	b.issuersLock.Lock()
+	defer b.issuersLock.Unlock()
+
+	// Fetch and parse our issuers so we have their expiration date.
+	sc := b.makeStorageContext(ctx, req.Storage)
+	issuerIDCertMap, err := fetchIssuerMapForRevocationChecking(sc)
+	if err != nil {
+		return err
+	}
+
+	// Fetch the issuer config to find the default; we don't want to remove
+	// the current active issuer automatically.
+	iConfig, err := sc.getIssuersConfig()
+	if err != nil {
+		return err
+	}
+
+	// We want certificates which have expired before this date by a given
+	// safety buffer. So we subtract the buffer from now, and anything which
+	// has expired before our after buffer can be tidied, and anything that
+	// expired after this buffer must be kept.
+	now := time.Now()
+	afterBuffer := now.Add(-1 * config.IssuerSafetyBuffer)
+
+	rebuildChainsAndCRL := false
+
+	for issuer, cert := range issuerIDCertMap {
+		if cert.NotAfter.After(afterBuffer) {
+			continue
+		}
+
+		entry, err := sc.fetchIssuerById(issuer)
+		if err != nil {
+			return nil
+		}
+
+		// This issuer's certificate has expired. We explicitly persist the
+		// key, but log both the certificate and the keyId to the
+		// informational logs so an admin can recover the removed cert if
+		// necessary or remove the key (and know which cert it belonged to),
+		// if desired.
+		msg := "[Tidy on mount: %v] Issuer %v has expired by %v and is being removed.\n"
```

Comment:
Since this is going out the logger, Rather than being a paragraph that's human readable, it may be better to use it's "key","value" paradigm so that it is machine parseable.  Possibly base64 the cert?

Reply:
Do we have precedence/an example for that somewhere? I'm not immediately seeing any... Would JSON be better?

Why would we base64 the already PEM-encoded certificate out of curiosity? Removing the line breaks within the PEM blob?

Reply:
yeah.  It may do that already.  The Logger() functions signature is (msg string, args ...interface{}), expecting pairs of subsequent args with the first being a key string and the second it's value.  Then the logger formats that differently depending on whether its text or json output (I think).

c.f. path_tidy.go L259

Reply:
Ah gotcha. I'll do that then.

Reply:
Yep, looks good.